support/Python/Machines.py

```diff
@@ -0,0 +1,74 @@
+# Distributed under the MIT License.
+# See LICENSE.txt for details.
+"""Support for host machines, such as supercomputers.
+
+Machines are defined as YAML files in 'support/Machines/'. To add support for a
+new machine, add a YAML file that defines a `Machine:` key with the attributes
+listed in the `Machine` class below.
+
+To select a machine, specify the `MACHINE` option when configuring the CMake
+build.
+"""
+
+import os
+import yaml
+from dataclasses import dataclass
+
+
+@dataclass(frozen=True)
+class Machine(yaml.YAMLObject):
+    """A machine we know how to run on, such as a particular supercomputer.
+
+    Attributes:
+      Name: A short name for the machine. Must match the YAML file name.
+      Description: A description of the machine. Give some basic context and
+        any information that may help people get started using the machine.
+        Provide links to wiki pages, signup pages, etc., for additional
+        information.
+      DefaultProcsPerNode: Default number of worker threads spawned per node.
+        It is often advised to leave one core per node or socket free for
+        communication, so this might be the number of cores or hyperthreads
+        per node minus one.
```

Comment:
Should we have this taken care of somewhere else? By that I mean, the machine file just specifies how many total cores per node there are, and then somewhere else (maybe when creating submit scripts later on) we specify how many comm/worker cores we want.

The reason I say this is because on the larger machines we have access to (Frontera, Anvil, ...) where there are a ton of cores I found for the BBHs that I actually needed multiple comm cores per physical node, effectively splitting the physical node up into several virtual charm nodes. However, the number of virtual nodes we want to create could be problem dependent (maybe splitting the node in half, maybe 3 virtual nodes, etc...), so in that case we wouldn't want that specified here in the machine file.

Thoughts?

Reply:
Sorry, I probably wasn't super clear about the purpose of these values. I added the following to the docstring:

> Many configuration options for job submission are hardcoded in the submit
> script for the machine (such as the total number of available cores per
> node). Here we provide parameters that we want to adjust when scheduling
> jobs, but that have sensible defaults based on our experience running on the
> machine. Parameters are set to 'None' when they have no sensible default for
> the machine.

Does that help? For the PPN we could delete the default and require users to always specify it. However, I think the default is useful because I don't want to remember every time when I schedule a run how many cores the machine has and what's the recommended number of worker threads on this machine. The default is based on experience. It can still be adjusted when scheduling a job if necessary, e.g. to do scaling tests. If the ideal PPN ends up being very problem-dependent then we can think about a way to set it based on the problem. For the majority of cases I don't want to think about this every time I schedule a run though.

Reply:
Yeah that makes it clearer. I guess we'll need a utility to edit submit scripts (and update their parameters) once we decide on how we want submit scripts to look.

Reply:
Yea I have that written already. My plan is to treat submit scripts as templates with placeholders that can be substituted. They essentially already are templates, only they are edited by hand. The syntax for placeholders is `{{ num_nodes }}`, which follows Jinja (https://jinja.palletsprojects.com/en/3.1.x/templates/). So you instead of the current `SPECTRE_BUILD_DIR=/path/to/build/dir` that's in many submit scripts we write `SPECTRE_BUILD_DIR={{ build_dir }}` and the parameter gets substituted.